### ISSUE TITLE
[问题/Issue] 章节2.2.6：DecoderLayer中forward部分逻辑问题 / Chapter2.2.6: Something maybe wrong in DecoderLayer`s forward()

### ISSUE THREAD

**Issue by ooooolong** [datawhalechina/happy-llm#21]
### 1. 遇到问题的章节 / Affected Chapter

2.2.6

### 2. 具体问题描述 / Problem Description

![Image](https://github.com/user-attachments/assets/447cf0c9-ead2-461e-b3b9-90595249af85)
掩码注意力部分和多头自注意力部分逻辑不太清楚是不是有点逻辑问题？先对x做归一化之后，不就把之后残差所用的原始x也给替换掉了吗？而残差的要求不是保留原始输入辅助后续学习的吗？
期待得到您的回复！谢谢

### 3. 问题重现材料 / Reproduction Materials

class DecoderLayer(nn.Module):
  '''解码层'''
    def __init__(self, args):
        super().__init__()
        # 一个 Layer 中有三个 LayerNorm，分别在 Mask Attention 之前、Self Attention 之前和 MLP 之前
        self.attention_norm_1 = LayerNorm(args.n_embd)
        # Decoder 的第一个部分是 Mask Attention，传入 is_causal=True
        self.mask_attention = MultiHeadAttention(args, is_causal=True)
        self.attention_norm_2 = LayerNorm(args.n_embd)
        # Decoder 的第二个部分是 类似于 Encoder 的 Attention，传入 is_causal=False
        self.attention = MultiHeadAttention(args, is_causal=False)
        self.ffn_norm = LayerNorm(args.n_embd)
        # 第三个部分是 MLP
        self.feed_forward = MLP(args)

    def forward(self, x, enc_out):
        **# Layer Norm
        x = self.attention_norm_1(x)
        # 掩码自注意力
        x = x + self.mask_attention.forward(x, x, x)
        # 多头注意力
        x = self.attention_norm_2(x)
        h = x + self.attention.forward(x, enc_out, enc_out)**
        # 经过前馈神经网络
        **out = h + self.feed_forward.forward(self.fnn_norm(h))**
        return out

### 确认事项 / Verification

- [x] 此问题未在过往Issue中被报告过 / This issue hasn't been reported before

**Comment by Blackteaxx** [comment]
Chapter5的代码是这样的

```python
class DecoderLayer(nn.Module):
    def __init__(self, layer_id: int, args: ModelConfig):
        super().__init__()
        # 定义多头注意力的头数
        self.n_heads = args.n_heads
        # 定义输入维度
        self.dim = args.dim
        # 定义每个头的维度，等于输入维度除以头数
        self.head_dim = args.dim // args.n_heads
        # 定义LLaMA2Attention对象，用于进行多头注意力计算
        self.attention = Attention(args)
        # 定义LLaMAMLP对象，用于进行前馈神经网络计算
        self.feed_forward = MLP(
            dim=args.dim,
            hidden_dim=args.hidden_dim,
            multiple_of=args.multiple_of,
            dropout=args.dropout,
        )
        # 定义层的ID
        self.layer_id = layer_id
        # 定义注意力计算的归一化层
        self.attention_norm = RMSNorm(args.dim, eps=args.norm_eps)
        # 定义前馈神经网络计算的归一化层
        self.ffn_norm = RMSNorm(args.dim, eps=args.norm_eps)

    def forward(self, x, freqs_cos, freqs_sin):
        # 前向传播函数
        # 首先，输入x经过注意力归一化层，然后进行注意力计算，结果与输入x相加得到h
        # 然后，h经过前馈神经网络归一化层，然后进行前馈神经网络计算，结果与h相加得到输出
        h = x + self.attention.forward(self.attention_norm(x), freqs_cos, freqs_sin)
        out = h + self.feed_forward.forward(self.ffn_norm(h))
        return out
```

确实这个 ch2 中的代码重新赋值存在一些问题

**Comment by ooooolong** [comment]
谢谢你的回答！



发自我的iPhone


------------------ 原始邮件 ------------------
From: Blackteaxx ***@***.***&gt;
发送时间: 06/08/2025, 16:34
To: datawhalechina/happy-llm ***@***.***&gt;
抄送: LaiD ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [datawhalechina/happy-llm] [问题/Issue] 章节2.2.6：DecoderLayer中forward部分逻辑问题 / Chapter2.2.6: Something maybe wrong in DecoderLayer`s forward() (Issue #21)



Blackteaxx left a comment (datawhalechina/happy-llm#21)
 
Chapter5的代码是这样的
 class DecoderLayer(nn.Module):     def __init__(self, layer_id: int, args: ModelConfig):         super().__init__()         # 定义多头注意力的头数         self.n_heads = args.n_heads         # 定义输入维度         self.dim = args.dim         # 定义每个头的维度，等于输入维度除以头数         self.head_dim = args.dim // args.n_heads         # 定义LLaMA2Attention对象，用于进行多头注意力计算         self.attention = Attention(args)         # 定义LLaMAMLP对象，用于进行前馈神经网络计算         self.feed_forward = MLP(             dim=args.dim,             hidden_dim=args.hidden_dim,             multiple_of=args.multiple_of,             dropout=args.dropout,         )         # 定义层的ID         self.layer_id = layer_id         # 定义注意力计算的归一化层         self.attention_norm = RMSNorm(args.dim, eps=args.norm_eps)         # 定义前馈神经网络计算的归一化层         self.ffn_norm = RMSNorm(args.dim, eps=args.norm_eps)     def forward(self, x, freqs_cos, freqs_sin):         # 前向传播函数         # 首先，输入x经过注意力归一化层，然后进行注意力计算，结果与输入x相加得到h         # 然后，h经过前馈神经网络归一化层，然后进行前馈神经网络计算，结果与h相加得到输出         h = x + self.attention.forward(self.attention_norm(x), freqs_cos, freqs_sin)         out = h + self.feed_forward.forward(self.ffn_norm(h))         return out
 
确实这个 ch2 中的代码重新赋值存在一些问题

**Comment by surpoloyang** [comment]
同样疑惑
